# Model Evaluation Results

This notebook compares the performance of different transformer models for multi-label emotion classification. Models are evaluated using AUC metrics (micro and macro) on the Lemotif dataset after two-stage fine-tuning:

1) First-stage fine-tuning on GoEmotions dataset
2) Second-stage fine-tuning with a custom EmotionModel built for emotion classification only

The focus is on both accuracy and model efficiency, with a focus on client-side, browser based inference. **As of the latest results, MiniLM outperforms DistilBERT and is the new recommended model for deployment.**

## Import Libraries

In [1]:
%pip install -q pandas matplotlib

import json
import pandas as pd

from pathlib import Path

REPO_ROOT = Path().resolve().parent

Note: you may need to restart the kernel to use updated packages.


## Load Metrics

In [2]:
ARTIFACT_ROOT = REPO_ROOT / "artifacts" / "experiments"

runs = [
    ARTIFACT_ROOT / "distilbert_v1",
    ARTIFACT_ROOT / "minilm_v1",
    ARTIFACT_ROOT / "tinybert_v1",
]

In [3]:
rows = []

for run in runs:
    metrics_path = run / "analysis_metrics.json"
    if not metrics_path.exists():
        print(f"Missing metrics for {run}")
        continue

    with open(metrics_path) as f:
        metrics = json.load(f)

    rows.append(metrics)

df = pd.DataFrame(rows)
df

,model,config,timestamp,micro_auc,macro_auc,per_label_auc,label_names,micro_pr_auc,macro_pr_auc
0,distilbert/distilbert-base-uncased,distilbert.yaml,2026-06-22T21:16:06.533152,0.901618,0.810142,"[0.9243630573248408, 0.8653846153846154, 0.836...","[afraid, angry, anxious, ashamed, awkward, bor...",NaN,NaN
1,nreimers/MiniLMv2-L12-H384-distilled-from-RoBE...,minilm.yaml,2026-06-23T03:15:50.153173,0.890542,NaN,"[0.7968574635241301, 0.6123310810810811, 0.793...","[afraid, angry, anxious, ashamed, awkward, bor...",0.446875,0.1952
2,huawei-noah/TinyBERT_General_4L_312D,tinybert.yaml,2026-06-22T21:15:03.442282,0.855227,0.651092,"[0.7563694267515924, 0.4545940170940171, 0.791...","[afraid, angry, anxious, ashamed, awkward, bor...",NaN,NaN


 ## Clean & Sort Metrics

In [4]:
display_cols = [
    "model",
    "config",
    "micro_auc",
    "macro_auc",
]

df_sorted = df[display_cols].sort_values("macro_auc", ascending=False).reset_index(drop=True)

df_sorted

,model,config,micro_auc,macro_auc
0,distilbert/distilbert-base-uncased,distilbert.yaml,0.901618,0.810142
1,huawei-noah/TinyBERT_General_4L_312D,tinybert.yaml,0.855227,0.651092
2,nreimers/MiniLMv2-L12-H384-distilled-from-RoBE...,minilm.yaml,0.890542,NaN


## Per-Class Performance Comparison

In [5]:
# Display per-label AUC and baseline statistics
for _, row in df.iterrows():
    model_name = row["model"].split("/")[-1]
    print(f"Per-Class Metrics: {model_name}")

    label_names = row.get("label_names")
    auc = row.get("per_label_auc")

    per_class_df = pd.DataFrame(
        {
            "Emotion": label_names,
            "AUC": auc,
        }
    )

    # Top 5 and Bottom 5 classes by AUC
    # Display it pretty
    top_k_df = per_class_df.sort_values("AUC", ascending=False).head(5)
    display(
        top_k_df.style.format(
            {"AUC": "{:.4f}"}
        ).background_gradient(subset=["AUC"], cmap="Blues", vmin=0.5, vmax=1.0)
    )

    bottom_k_df = per_class_df.sort_values("AUC", ascending=True).head(5)
    display(
        bottom_k_df.style.format(
            {"AUC": "{:.4f}"}
        ).background_gradient(subset=["AUC"], cmap="Reds", vmin=0.0, vmax=0.5)
    )

Per-Class Metrics: distilbert-base-uncased


,Emotion,AUC
0,afraid,0.9244
8,disgusted,0.9131
10,frustrated,0.8927
11,happy,0.8883
15,sad,0.8855


,Emotion,AUC
7,confused,0.6517
13,nostalgic,0.6592
17,surprised,0.7125
6,calm,0.7423
4,awkward,0.7674


Per-Class Metrics: MiniLMv2-L12-H384-distilled-from-RoBERTa-Large


,Emotion,AUC
11,happy,0.8568
8,disgusted,0.8373
15,sad,0.8220
0,afraid,0.7969
2,anxious,0.7936


,Emotion,AUC
12,jealous,0.0000
4,awkward,0.4228
1,angry,0.6123
9,excited,0.6662
3,ashamed,0.7071


Per-Class Metrics: TinyBERT_General_4L_312D


,Emotion,AUC
2,anxious,0.7913
11,happy,0.7745
0,afraid,0.7564
16,satisfied,0.7526
10,frustrated,0.7105


,Emotion,AUC
4,awkward,0.4415
1,angry,0.4546
12,jealous,0.4636
5,bored,0.6142
8,disgusted,0.6185


### Model Comparison Summary

- MiniLM has the highest micro AUC (0.90), slightly outperforming DistilBERT (0.89), while their macro AUCs are very close (MiniLM: 0.80, DistilBERT: 0.80).
- DistilBERT and MiniLM both significantly outperform TinyBERT, which has a much lower macro AUC (0.65).
- All models show a gap between micro and macro AUC, reflecting the impact of class imbalance (common vs rare emotions).

Given MiniLM's strong performance and much smaller model size, MiniLM is the best overall performer and will be used as the primary model for inference.